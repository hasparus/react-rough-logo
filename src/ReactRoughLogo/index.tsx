import React from "react";
import rough from "roughjs";

import "./styles.css";
import { RoughSVG } from "roughjs/bin/svg";

// https://github.com/pshihn/rough/wiki
type RoughOptions = {
  roughness?: number;
  bowing?: number;
  stroke?: string;
  strokeWidth?: number;
  fill?: string;
  fillStyle?:
    | "hachure"
    | "solid"
    | "zigzag"
    | "cross-hatch"
    | "dots";
  fillWeight?: number;
  hachureAngle?: number;
  hachureGap?: number;
  curveStepCount?: number;
  simplification?: number;
};

const defaultConfig = {
  scale: 1.5,
  ellipsesOptions: { roughness: 2 } as RoughOptions,
  circleOptions: {} as RoughOptions,
};

type Config = typeof defaultConfig;

function mountShapes(
  svgElement: SVGSVGElement,
  { circleOptions, ellipsesOptions, scale }: Config
) {
  svgElement.innerHTML = "";
  const rc = rough.svg(svgElement) as RoughSVG;
  for (let i = 0; i < 3; i++) {
    svgElement.appendChild(
      rc.ellipse(
        0,
        0,
        scale * 570,
        scale * 220,
        ellipsesOptions
      )
    );
  }
  svgElement.appendChild(
    rc.circle(0, 0, scale * 100, circleOptions)
  );
}

type Props = typeof defaultConfig &
  React.SVGProps<SVGSVGElement>;

class ReactRoughLogo extends React.PureComponent<Props> {
  static displayName =
    process.env.NODE_ENV === "production"
      ? null
      : "RoughReactLogo";
  static defaultProps = defaultConfig;

  svg = React.createRef<SVGSVGElement>();

  _mountShapes() {
    if (this.svg.current) {
      mountShapes(this.svg.current, this.props);
    }
  }

  componentDidMount() {
    this._mountShapes();
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps !== this.props) {
      this._mountShapes();
    }
  }

  render() {
    const {
      style,
      scale,
      ellipsesOptions,
      circleOptions,
      ...rest
    } = this.props;
    return (
      <svg
        className={"react-rough-logo"}
        ref={this.svg}
        viewBox="-500 -500 1000 1000"
        style={{
          height: "100%",
          width: "100%",
          ...style,
        }}
        {...rest}
      />
    );
  }
}

export default ReactRoughLogo;
