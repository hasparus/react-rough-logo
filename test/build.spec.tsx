import React from "react";
import { shallow, render } from "enzyme";

import ReactRoughLogo from "../build/index";

import EsmLogo from "../build/index.esm";

describe("Built index.js", () => {
  describe("ReactRoughLogo", () => {
    it("matches snapshot", () => {
      expect(render(<ReactRoughLogo />)).toMatchSnapshot();
    });

    it("renders svg", () => {
      const logo = shallow(<ReactRoughLogo />);
      expect(logo.find("svg").exists()).toBe(true);
    });
  });
});

describe("Built index.esm.js", () => {
  const Logo = (EsmLogo as any) as typeof ReactRoughLogo; // FIXME
  describe("ReactRoughLogo", () => {
    it("matches snapshot", () => {
      expect(render(<Logo />)).toMatchSnapshot();
    });

    it("renders svg", () => {
      const logo = shallow(<Logo />);
      expect(logo.find("svg").exists()).toBe(true);
    });
  });
});
