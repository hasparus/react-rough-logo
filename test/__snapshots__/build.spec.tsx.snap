// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Built index.esm.js ReactRoughLogo matches snapshot 1`] = `
<svg
  class="react-rough-logo"
  config="[object Object]"
  scale="1.5"
  style="height:100%;width:100%"
  viewBox="-500 -500 1000 1000"
/>
`;

exports[`Built index.js ReactRoughLogo matches snapshot 1`] = `
<svg
  class="react-rough-logo"
  config="[object Object]"
  scale="1.5"
  style="height:100%;width:100%"
  viewBox="-500 -500 1000 1000"
/>
`;
